File: WebClient SDK/src/default/workflow/krProcess/krLocalTileCommand.ts
import { IKrTileCommand } from './interfaces';
import { IUIContext, showConfirm } from 'tessa/ui';
import {
  KrProcessInstance,
  KrTileInfo,
  launchProcessWithCardEditor
} from 'tessa/workflow/krProcess';
import { Guid } from 'tessa/platform';
import { ICardModel, CardSavingRequest, CardSavingMode } from 'tessa/ui/cards';
import { ITile } from 'tessa/ui/tiles';

export class KrLocalTileCommand implements IKrTileCommand {
  //#region ctor

  private constructor() {}

  //#endregion

  //#region instance

  private static _instance: KrLocalTileCommand;

  public static get instance(): KrLocalTileCommand {
    if (!KrLocalTileCommand._instance) {
      KrLocalTileCommand._instance = new KrLocalTileCommand();
    }
    return KrLocalTileCommand._instance;
  }

  //#endregion

  public async onClickAction(context: IUIContext, _tile: ITile, tileInfo: KrTileInfo) {
    const cardEditor = context.cardEditor;
    let model: ICardModel;
    if (tileInfo.id === Guid.empty || !cardEditor || !(model = cardEditor.cardModel!)) {
      return;
    }

    if (tileInfo.askConfirmation) {
      const result = await showConfirm(tileInfo.confirmationMessage);
      if (!result) {
        return;
      }
    }

    await cardEditor.setOperationInProgress(async () => {
      if (await model.hasChanges()) {
        if (
          !(await cardEditor.saveCard(
            context,
            undefined,
            new CardSavingRequest(CardSavingMode.RefreshOnSuccess)
          ))
        ) {
          return;
        }
      }
      const process = KrProcessInstance.createWithParams({
        processId: tileInfo.id,
        processInfo: {},
        cardId: context.cardEditor!.cardModel!.card.id
      });

      await launchProcessWithCardEditor(process, cardEditor, true);
    });
  }
}

File: Source/Extensions/Tessa.Extensions.Default.Client/Workflow/KrProcess/KrLocalTileCommand.cs
using System.Threading.Tasks;
using Tessa.Extensions.Default.Shared.Workflow.KrProcess;
using Tessa.Localization;
using Tessa.UI;
using Tessa.UI.Cards;
using Tessa.UI.Tiles;

namespace Tessa.Extensions.Default.Client.Workflow.KrProcess
{
    /// <summary>
    /// Команда выполняющаяся при клике по локальному тайлу подсистемы маршрутов.
    /// </summary>
    public sealed class KrLocalTileCommand : IKrTileCommand
    {
        private readonly IKrProcessLauncher launcher;

        public KrLocalTileCommand(
            IKrProcessLauncher launcher)
        {
            this.launcher = launcher;
        }

        /// <inheritdoc />
        public async Task OnClickAsync(
            IUIContext context,
            ITile tile,
            KrTileInfo tileInfo)
        {
            var cardEditor = context?.CardEditor;
            if (tileInfo.ID == default
                || cardEditor == null)
            {
                return;
            }

            if (tileInfo.AskConfirmation
                && !TessaDialog.Confirm(LocalizationManager.Format(tileInfo.ConfirmationMessage)))
            {
                return;
            }

            using (TessaSplash.Create("$KrButton_DefaultTileSplash"))
            using (cardEditor.SetOperationInProgress(blocking: true))
            {
                // Двойное сохранение, если в карточке есть изменения. Сперва делается основное сохранение, затем запуск процесса.
                if (await cardEditor.CardModel.HasChangesAsync())
                {
                    if (!await cardEditor.SaveCardAsync(
                            context,
                            request: new CardSavingRequest(CardSavingMode.RefreshOnSuccess)))
                    {
                        return;
                    }
                }

                var process = KrProcessBuilder
                    .CreateProcess()
                    .SetProcess(tileInfo.ID)
                    .SetCard(context.CardEditor.CardModel.Card.ID)
                    .Build();

                await this.launcher.LaunchWithCardEditorAsync(process, cardEditor, true);
            }
        }
    }
}